File: Intersection of Two Arrays II.java
Approach:
Iterate over the first array(smaller array) and search for thr element in the 2nd array(first occurance only)
if the element is found in 2nd array update the low index to mid+1 so we do not check the same element again

TC : m log n
m = lower size array
n = larger array

SC : O(1)

class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        int n1 = nums1.length;
        int n2 = nums2.length;
        if(n1 > n2)
            return intersect(nums2,nums1);
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        List<Integer> res = new ArrayList<>();
        int low = 0;
        for(int i=0;i<n1;i++){
            int index = firstIndexSearch(nums1[i],nums2,low, n2-1);
            if(index != -1){
                low = index + 1;
                res.add(nums1[i]);
            }
        }

        int [] ans = new int[res.size()];
        for(int i =0; i < res.size();++i)
            ans[i] = res.get(i);

        return ans;
    }

    private int firstIndexSearch(int target , int[] arr, int low, int high){   
        while(low <= high){
            int mid = low + (high - low)/2;
            if(arr[mid] == target){
                if(mid == low || arr[mid-1] != arr[mid])
                    return mid;
                else
                    high = mid -1;
            } else if(arr[mid] > target){
                high = mid-1;
            } else{
                low = mid + 1;
            }
        }
        return -1;
    }
}

File: Intersection of Two Arrays II.cpp
TC : n1 + n2 
  n1 - size of nums1
  n2 - size of nums2

SC : Min(n1,n2)

class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        int n1 = nums1.size();
        int n2 = nums2.size();
        if(n1 > n2)
            intersect(nums2,nums1);
        unordered_map<int,int> mp;
        vector<int> ans;
        for(int i =0 ;i< n1;++i)
            mp[nums1[i]] += 1;
        
        for(int i=0 ; i< n2 ; ++i){
            if(mp[nums2[i]]){
               mp[nums2[i]] -= 1;
                ans.emplace_back(nums2[i]);
            }
        }
        return ans;
    }
};
